equinox/nn/_batch_norm.py

```diff
@@ -40,25 +41,69 @@ class BatchNorm(StatefulLayer, strict=True):
     statistics updated. During inference then just the running statistics are used.
     Whether the model is in training or inference mode should be toggled using
     [`equinox.nn.inference_mode`][].
+
+    With `mode = "batch"` during training the batch mean and variance are used
+    for normalization.  For inference the exponential running mean and ubiased
+    variance are used for normalization. This is in line with out other JAX
+    packages (e.g. haiku, flax) implement batch norm.
+
+    With `mode = "ema"` exponential running means and variances are kept.  During
+    training the batch statistics are used to fill in the running statistics until
+    they are populated.  During inference the running statistics are used for
+    normalization.
+
+    ??? cite
+
+        [Batch Normalization: Accelerating Deep Network Training by Reducing
+         Internal Covariate Shift](https://arxiv.org/abs/1502.03167)
+
+        ```bibtex
+        @article{DBLP:journals/corr/IoffeS15,
+        author       = {Sergey Ioffe and
+                        Christian Szegedy},
+        title        = {Batch Normalization: Accelerating Deep Network Training
+                        by Reducing Internal Covariate Shift},
+        journal      = {CoRR},
+        volume       = {abs/1502.03167},
+        year         = {2015},
+        url          = {http://arxiv.org/abs/1502.03167},
+        eprinttype    = {arXiv},
+        eprint       = {1502.03167},
+        timestamp    = {Mon, 13 Aug 2018 16:47:06 +0200},
+        biburl       = {https://dblp.org/rec/journals/corr/IoffeS15.bib},
+        bibsource    = {dblp computer science bibliography, https://dblp.org}
+        }
+        ```
     """  # noqa: E501
 
     weight: Float[Array, "input_size"] | None
     bias: Float[Array, "input_size"] | None
-    first_time_index: StateIndex[Bool[Array, ""]]
-    state_index: StateIndex[
-        tuple[Float[Array, "input_size"], Float[Array, "input_size"]]
+    ema_first_time_index: Optional[StateIndex[Bool[Array, ""]]]
+    ema_state_index: Optional[
+        StateIndex[tuple[Float[Array, "input_size"], Float[Array, "input_size"]]]
+    ]
+    batch_counter: Optional[StateIndex[Int[Array, ""]]]
+    batch_state_index: Optional[
+        StateIndex[
+            tuple[
+                tuple[Float[Array, "input_size"], Float[Array, "input_size"]],
+                tuple[Float[Array, "input_size"], Float[Array, "input_size"]],
+            ],
+        ]
     ]
     axis_name: Hashable | Sequence[Hashable]
     inference: bool
     input_size: int = field(static=True)
     eps: float = field(static=True)
     channelwise_affine: bool = field(static=True)
     momentum: float = field(static=True)
+    mode: Literal["ema", "batch"] = field(static=True)
 
     def __init__(
         self,
         input_size: int,
         axis_name: Hashable | Sequence[Hashable],
+        mode: str = "ema",
```

Comment:
Should we make the default `"legacy"` -- which emits a warning and then sets `"ema"` -- to encourage picking explicitly?


Also the type annotation here could be a Literal.